src/librustc/mir/interpret/allocation.rs

```diff
@@ -192,21 +193,16 @@ impl<'tcx, Tag: Copy, Extra: AllocationExtra<Tag>> Allocation<Tag, Extra> {
     /// Returns the range of this allocation that was meant.
     #[inline]
     fn check_bounds(&self, offset: Size, size: Size) -> Range<usize> {
-        let end = offset + size; // This does overflow checking.
-        assert_eq!(
-            end.bytes() as usize as u64,
-            end.bytes(),
-            "cannot handle this access on this host architecture"
-        );
-        let end = end.bytes() as usize;
+        let end = Size::add(offset, size); // This does overflow checking.
```

Comment:
I don't like the `Size::bin_op` style very much. Why are the regular binary operators not ok?

Reply:
When I see a `+` I have to assume it is not overflow-checked. Figuring out the type of the operands can be non-trivial. So the only way to write overflow-clean Rust code for sure is to make sure there is, syntactically, no `+`/`*`/*-` anywhere in the code. Hence the explicit `Size::{add,mul}`.

Reply:
The alternative is to just turn on overflow checks in release mode.

I still don't like it. This is much less readable and most of the time the types are a line or two above in the arguments.

Reply:
I'm not gonna block the PR on it, but I want someone else to also say it's ok to write lisp-math in rustc if infix math would produce the exact same machine code.

Reply:
Now that I did the audit, I think I'm fine going back to infix notation. I will add a comment though in places where the type is not immediately obvious.

> The alternative is to just turn on overflow checks in release mode.

Would be great if that could be controlled on a per-module basis.

Reply:
we could probably start out with a `rustc_overflow_checks` attribute, seems straight forward enough to do.

Reply:
> we could probably start out with a rustc_overflow_checks attribute, seems straight forward enough to do.

https://github.com/rust-lang/rust/issues/70358